### DISCUSSION TITLE
Importing from macOS misses URLs

### DISCUSSION THREAD

**Discussion by SuperKenVery** [dani-garcia/vaultwarden#2707] [Q&A] [answered]
### Subject of the issue

When importing from macOS, all URLs are gone. 

### Deployment environment

* vaultwarden version: 1.23.1

* Install method: docker image

* Clients used: web, safari

* Reverse proxy and version: caddy 2.5.2

* MySQL/MariaDB or PostgreSQL version: 

* Other relevant details:

### Steps to reproduce

Export from macOS `Preference` app. 
Import on web

Import succeeds, but all URLs are gone.

**Comment by Masgalor** [comment]
Hi @SuperKenVery .
Imports and exports are done by the clients, vaultwarden is not involved.
Did you check if this a known problem with your clients?
Did you check if the exported data includes the URLs?

You did not mention what version of the webvault you are running, but considering the version of vaultwarden (which is really dated) you are likely running a very old webvault. Maybe the export from the recent app could not be understood by an older webvault. You should update both, vaultwarden an your webvault.

**Reply by SuperKenVery** [reply]
Sorry but I can't get what you are talking about. 

I installed `vaultwarden` using docker image (as an alternative for `Bitwarden`) to self-host a `Bitwarden` server, and accessed it using a browser. 

> Imports and exports are done by the clients, vaultwarden is not involved.

What is `the clients`? The only things I used are `vault warden` and a browser. 

> Did you check if the exported data includes the URLs?

I checked, they include URLs. 

> but considering the version of vaultwarden (which is really dated) you are likely running a very old webvault

I pulled the image from docker hub just a few days ago...

And, what is web vault? I understand it as vault warden in browser. Then it shouldn't have its own version; it is related to vault warden.

**Reply by BlackDex** [reply]
> Sorry but I can't get what you are talking about.
> 
> I installed `vaultwarden` using docker image (as an alternative for `Bitwarden`) to self-host a `Bitwarden` server, and accessed it using a browser.

Vaultwarden uses the Bitwarden web-vault with only some very small changes to make it work correctly with Vaultwarden.
All logic etc... is not touched. So the client is the actual web client provided by Bitwarden.

> 
> > Imports and exports are done by the clients, vaultwarden is not involved.
> 
> What is `the clients`? The only things I used are `vault warden` and a browser.
> 

The web-vault is also a client, just like the mobile app, or the browser extension.

> > Did you check if the exported data includes the URLs?
> 
> I checked, they include URLs.
> 
> > but considering the version of vaultwarden (which is really dated) you are likely running a very old webvault
> 
> I pulled the image from docker hub just a few days ago...
> 
> And, what is web vault? I understand it as vault warden in browser. Then it shouldn't have its own version; it is related to vault warden.

The version you are running is from Dec 14, 2021, so that is very old.
The latest version currently available is 1.25.2. You probably are using `bitwarden_rs` (And even that has 1.24) instead of `vaultwarden`
Please use: https://hub.docker.com/r/vaultwarden/server and make sure you pull, stop, rm, start the container. Not removing the container will not update the image.

**Reply by Masgalor** [reply]
No problem, we will sort it out.

Vaultwarden is a backend service that provides an API compatible with bitwarden.
You can not use it directly, you need a client to make use of it.
What you see in your browser is a client called web-vault, what you use on MacOS is a client called BitwardenDesktop, what you use on your phone is a client called BitwardenMobile. Vaultwarden only provides an endpoint for those clients to interact with, they are not developed by vaultwarden. All clients are developed by bitwarden https://github.com/bitwarden/clients

The most recent version of vaultwarden is `1.25.2` you are running `1.23.1`.
I dont use docker, but `docker pull vaultwarden/server:latest` downloads `1.25.2` for me.

The official docker images contain a compatible web-vault version, no worries.
I did assume you are something customzied because of the old vaultwarden version.
I cant help with docker, but at first you should find out why it installs the wrong version, you should have vaultwarden `1.25.2` and webvault `2022.6.2`.

**Reply by SuperKenVery** [reply]
Thanks for all your kind reply! 

@Masgalor I tried to pull the image (vaultwarden/server:latest), but got a message that mine was up to date. I'm using M1, by the way. Don't know if that influences. 

> Imports and exports are done by the clients, vaultwarden is not involved.

You mean the files are parsed by clients, then they request vault warden to store them? Then should I file an issue in web vault's repository instead?

**Reply by BlackDex** [reply]
@SuperKenVery 
I suggest to carefully read this wiki page on how to update the container: https://github.com/dani-garcia/vaultwarden/wiki/Updating-the-vaultwarden-image

Or try to use something like docker-compose: https://github.com/dani-garcia/vaultwarden/wiki/Using-Docker-Compose
With compose you are able to simply run `docker compose pull ; docker compose up -d` which will trigger a restart/re-create when there is a new image available.

**Reply by SuperKenVery** [reply]
I admit that I don't know much about docker...

But if `docker pull` tells me that my image was up to date, I don't think anything like `docker-compose` helps. The pull command I used is the same as that in the wiki. 

Again, I suspect that the **arm64** version of the image doesn't point to the right image?

**Reply by BlackDex** [reply]
A `docker pull` only pulls the image. It will not magically update a running container, not even if you stop the container and then start it again.
You have to remove (rm) the container, and create it again, else it will keep using the previous image pulled when you created that container.
This is also explained here: https://github.com/dani-garcia/vaultwarden/wiki/Updating-the-vaultwarden-image

**Reply by SuperKenVery** [reply]
Tried what you said. Removed and re-created the container, but with no luck. 

`docker pull ...` reports
```
➜  ~ docker pull vaultwarden/server:latest
latest: Pulling from vaultwarden/server
Digest: sha256:08d8edab2311ee36de9f1e2dd7f6ae18680a7fcbf00bab98d63fd4ff82b285e2
Status: Image is up to date for vaultwarden/server:latest
docker.io/vaultwarden/server:latest
```

It still reports
```
/--------------------------------------------------------------------\
|                        Starting Vaultwarden                        |
|                           Version 1.23.1                           |
|--------------------------------------------------------------------|
```

**Reply by SuperKenVery** [reply]
Ohhh I think I know the problem. 
I'm using a mirror from `aliyun.com`

**Reply by BlackDex** [reply]
No, it is stating it pulled it from docker.io. You really need to do a `docker rm vaultwarden` and then a `docker run .....`
https://www.howtogeek.com/devops/how-to-upgrade-docker-containers-to-apply-image-updates/

Though, it doesn't reflect in the commands, that specific digest is indeed a very old version.
If that mirror acts as if it is docker.io, then i suggest to remove that mirror.

**Reply by SuperKenVery** [reply]
I'm pretty sure that it's pulling from the mirror from my packet capture results. Maybe `docker.io` is the user name. 

I did `docker rm vaultwaarden` and `docker run`, as I stated before. 

I'm in China mainland, and docker.io is very slow here. 😭 

Thanks again for all your help!

**Comment by SuperKenVery** [comment, marked as answer]
I've found the problem! By reading the web client's source code. It's a problem of the web client. 

The key name in the exported file is `URL`, but the client reads `Url`. 

However, this indeed has been fixed in the most recent version of the web client.  

Thanks for all your help!
